src/main/kotlin/uk/gov/dwp/dataworks/dks/services/impl/HttpKeyService.kt

```diff
@@ -0,0 +1,111 @@
+            }
+        }
+    }
+
+    override fun decryptKey(encryptionKeyId: String, encryptedKey: String): String {
+        logger.info("Decrypting encryptedKey: '$encryptedKey', keyEncryptionKeyId: '$encryptionKeyId'.")
+        try {
+            val cacheKey = "$encryptedKey/$encryptionKeyId"
+            return if (decryptedKeyCache.containsKey(cacheKey)) {
+                decryptedKeyCache[cacheKey]!!
+            }
+            else {
+                httpClientProvider.client().use { client ->
+                    val dksUrl = """$dataKeyServiceUrl/datakey/actions/decrypt?keyId=${URLEncoder.encode(encryptionKeyId, "US-ASCII")}"""
+                    logger.info("Calling dataKeyServiceUrl: '$dksUrl'.")
+                    val httpPost = HttpPost(dksUrl)
+                    httpPost.entity = StringEntity(encryptedKey, ContentType.TEXT_PLAIN)
+                    client.execute(httpPost).use { response ->
+                        val statusCode = response.statusLine.statusCode
+                        logger.info("dataKeyServiceUrl: '$dksUrl' returned status code '$statusCode'.")
+                        return when (statusCode) {
+                            200 -> {
+                                val entity = response.entity
+                                val text = BufferedReader(InputStreamReader(response.entity.content)).use(BufferedReader::readText)
+                                EntityUtils.consume(entity)
+                                val dataKeyResult = Gson().fromJson(text, DataKeyResult::class.java)
+                                decryptedKeyCache[cacheKey] = dataKeyResult.plaintextDataKey
+                                dataKeyResult.plaintextDataKey
+                            }
+                            400 ->
+                                throw DataKeyDecryptionException(
+                                    "Decrypting encryptedKey: '$encryptedKey' with keyEncryptionKeyId: '$encryptionKeyId' data key service returned status code '$statusCode'")
+                            else ->
+                                throw DataKeyServiceUnavailableException(
+                                    "Decrypting encryptedKey: '$encryptedKey' with keyEncryptionKeyId: '$encryptionKeyId' data key service returned status code '$statusCode'")
+                        }
+                    }
+                }
+            }
+        }
+        catch (ex: Exception) {
+            when (ex) {
+                is DataKeyDecryptionException, is DataKeyServiceUnavailableException -> {
+                    throw ex
+                }
+                else -> throw DataKeyServiceUnavailableException("Error contacting data key service: $ex")
+            }
+        }
+    }
+
+    fun clearCache() {
+        this.decryptedKeyCache = mutableMapOf()
+    }
+
+    private var decryptedKeyCache = mutableMapOf<String, String>()
```

Comment:
I think we have a memory leak here - if the cache is never cleared (or keys time out) then it will grow forever? I got round this using Google's `google.common.cache` in the previous implementation if that seems useful?